libbeat/common/file/interval_rotator.go

```diff
@@ -127,14 +106,53 @@ func (r *intervalRotator) LogPrefix(filename string, modTime time.Time) string {
 	return fmt.Sprintf("%s-%s-", filename, t.Format(r.fileFormat))
 }
 
-func (r *intervalRotator) NewInterval() bool {
-	now := r.clock.Now()
-	newInterval := r.newInterval(r.lastRotate, now)
-	return newInterval
+func (r *intervalRotator) RotatedFiles() []string {
+	files, err := filepath.Glob(r.filename + "*")
+	if err != nil {
+		if r.log != nil {
+			r.log.Debugw("failed to list existing logs: %+v", err)
+		}
+	}
+	r.SortIntervalLogs(files)
+	return files
 }
 
-func (r *intervalRotator) Rotate() {
-	r.lastRotate = r.clock.Now()
+func (r *intervalRotator) Rotate(reason rotateReason, t time.Time) error {
+	fi, err := os.Stat(r.ActiveFile())
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "failed to rotate backups")
+	}
+
+	logPrefix := r.LogPrefix(r.ActiveFile(), fi.ModTime())
+	files, err := filepath.Glob(logPrefix + "*")
+	if err != nil {
+		return errors.Wrap(err, "failed to list logs during rotation")
+	}
+
+	var targetFilename string
+	if len(files) == 0 {
+		targetFilename = logPrefix + "1"
+	} else {
+		r.SortIntervalLogs(files)
+		lastLogIndex, _, err := IntervalLogIndex(files[len(files)-1])
+		if err != nil {
+			return errors.Wrap(err, "failed to locate last log index during rotation")
+		}
+		targetFilename = logPrefix + strconv.Itoa(int(lastLogIndex)+1)
+	}
+
+	if err := os.Rename(r.ActiveFile(), targetFilename); err != nil {
```

Comment:
should we use SafeFileRotate here?

Reply:
It is unnecessary to create a temp file. If renaming fails we do not lose any data and we do not go into an inconsistent state, only the rotation stops. Stopping the rotation is not a problem, because it is going to be retried on the next `Write` call of the logger.

Furthermore, we had no complaints about using `os.Rename` during log rotation so far.